### PR TITLE
Added support for lame duck mode

### PR DESCRIPTION
**Pull request by kozlovic**
When receiving SIGUSR2 signal (or -sl ldm) the server stops
accepting new clients, closes routes connections and spread the
closing of client connections based on a config lame duck duration
(default is 30sec). This will help preventing a storm of client
reconnect when a server needs to be shutdown.

Signed-off-by: Ivan Kozlovic <ivan@synadia.com>

### CODE REVIEW

File: server/monitor.go [hunk 1]
```diff
@@ -1038,6 +1038,8 @@ func (reason ClosedState) String() string {
 		return "Route Removed"
 	case ServerShutdown:
 		return "Server Shutdown"
+	case LameDuckMode:
```
Comment (derekcollison) [inline server/monitor.go]: Agree now just use server shutdown as reason. No need for new one.

File: server/route.go [hunk 2]
```diff
@@ -1139,6 +1139,10 @@ func (s *Server) routeAcceptLoop(ch chan struct{}) {
 	for s.isRunning() {
 		conn, err := l.Accept()
 		if err != nil {
+			if s.isLameDuckMode() {
```
Comment (derekcollison) [inline server/route.go]: Why is this here?

File: server/server.go [hunk 3]
```diff
@@ -594,6 +599,15 @@ func (s *Server) AcceptLoop(clr chan struct{}) {
 	for s.isRunning() {
 		conn, err := l.Accept()
 		if err != nil {
+			if s.isLameDuckMode() {
```
Comment (derekcollison) [inline server/server.go]: Why would we not shut down the listeners by closing the socket vs this approach?

Comment (derekcollison) [reply server/server.go]: ok I see you use this to catch the state when you close the socket.

File: server/server.go [hunk 4]
```diff
@@ -1540,3 +1554,98 @@ func (s *Server) serviceListeners() []net.Listener {
...
+	for i := 0; i < expected; i++ {
+		<-s.ldmCh
+	}
+
+	durInMs := int64(s.getOpts().LameDuckDuration / time.Millisecond)
+	if durInMs < 1 {
+		durInMs = 1
+	}
+
+	// Close routes
```
Comment (derekcollison) [inline server/server.go]: Feels like this should not happen at all while we have connected clients. Maybe leave completely in place until real shutdown?

File: server/server.go [hunk 5]
```diff
@@ -1540,3 +1554,98 @@ func (s *Server) serviceListeners() []net.Listener {
...
+	s.mu.Lock()
+	routes := make([]*client, 0, len(s.routes))
+	for _, r := range s.routes {
+		routes = append(routes, r)
+	}
+	s.mu.Unlock()
+	for _, r := range routes {
+		r.setRouteNoReconnectOnClose()
+		r.closeConnection(LameDuckMode)
```
Comment (derekcollison) [inline server/server.go]: I think this is not what we want since it means clients who are still connected and communicating won't work.

File: server/server.go [hunk 6]
```diff
@@ -1540,3 +1554,98 @@ func (s *Server) serviceListeners() []net.Listener {
...
+	i := 0
+	for _, client := range s.clients {
+		clients[i] = client
+		i++
+	}
+	s.mu.Unlock()
+
+	s.Noticef("Closing existing clients")
+	for i, client := range clients {
+		client.closeConnection(LameDuckMode)
```
Comment (derekcollison) [inline server/server.go]: I think server shutdown ok, but can go either way.

File: server/server.go [hunk 7]
```diff
@@ -1540,3 +1554,98 @@ func (s *Server) serviceListeners() []net.Listener {
...
+		if sd > 0 || (i > 0 && i%pa == 0) {
+			time.Sleep(time.Duration(rand.Int63n(sd)))
+		}
+	}
+
+	s.Noticef("Lame duck mode ending")
```
Comment (derekcollison) [inline server/server.go]: Probably do not need this, shutdown will log.

File: server/server.go [hunk 8]
```diff
@@ -1540,3 +1554,98 @@ func (s *Server) serviceListeners() []net.Listener {
...
+	}
+	s.mu.Unlock()
+	for _, r := range routes {
+		r.setRouteNoReconnectOnClose()
+		r.closeConnection(LameDuckMode)
+	}
+
+	// Now capture all clients
+	s.mu.Lock()
+	numClients := int64(len(s.clients))
```
Comment (derekcollison) [inline server/server.go]: I would try to simplify. I would fire a timer every second and pre calculate batch size of how many to close per timer fire. Once zero your done.

File: server/monitor.go (hunk 1 above)
Comment (kozlovic) [reply server/monitor.go]: Ok, can revert this change

File: server/server.go (hunk 3 above)
Comment (kozlovic) [reply server/server.go]: When closing the listener, we would get an error. If we don't check for that and since server isRunning() is still true, we would go back in an accept, and fail again, etc.. More over, you don't want AcceptLoop to return right away because otherwise Start() will exit which means that the process will quit before we had a chance to do anything.

File: server/server.go (hunk 4 above)
Comment (kozlovic) [reply server/server.go]: If you don't close the route, then when a client connection is closed and it reconnects to another server, then that server sends its SUB back to the server in lame duck mode. That seem crazy to me. (noticed that when I was running with -DV)

File: server/server.go (hunk 5 above)
Comment (kozlovic) [reply server/server.go]: If you are ok with the fact that when client reconnect to another server this server starts to receive route SUBs, then I can remove that.

File: server/server.go (hunk 6 above)
Comment (kozlovic) [reply server/server.go]: Shutdown is indeed sufficient I think, because this is what is happening.

File: server/server.go (hunk 7 above)
Comment (kozlovic) [reply server/server.go]: Sure, will remove.

File: server/server.go (hunk 8 above)
Comment (kozlovic) [reply server/server.go]: Will think about it. Thanks!

File: server/server.go (hunk 3 above)
Comment (derekcollison) [reply server/server.go]: Gotcha..

File: server/server.go (hunk 4 above)
Comment (derekcollison) [reply server/server.go]: Possibly, but if we do this the clients that remain on this server can not communicate with remote subscribers, etc. And vice versa.

File: server/server.go (hunk 5 above)
Comment (derekcollison) [reply server/server.go]: I think for now that is ok. Say for example we have simple pub and sub connected to LD server, we migrate the subscriber. We want the PUB to succeed as best as possible. What might be good though if possible is to have the LD server let all the other servers to not advertise it as an option for reconnect.

File: server/server.go (hunk 8 above)
Comment (derekcollison) [reply server/server.go]: Sounds good, thanks for doing this. I think this will be a huge help to NGS.

File: server/server.go (hunk 4 above)
Comment (kozlovic) [reply server/server.go]: Ok, removed closing of routes.

File: server/server.go (hunk 5 above)
Comment (kozlovic) [reply server/server.go]: Removed closing of routes.

Comment (kozlovic) [reply server/server.go]: Also, I think we don't need complexity of communicating between server that server is in LD mode. Sine the LD server will have shutdown its listener port, even if client try to reconnect to it, they will fail and move to other server.

File: server/server.go (hunk 8 above)
Comment (kozlovic) [reply server/server.go]: The thing is that for test, I don't want to have it take so much time, this is why I was doing it this way where code still "works" even with very low duration (in the ms range). Still, will think about it more.

Comment (derekcollison) [reply server/server.go]: You could configure the timer interval for sure.

File: server/server.go [hunk 9]
```diff
@@ -1540,3 +1554,98 @@ func (s *Server) serviceListeners() []net.Listener {
...
+	defer s.mu.Unlock()
+	return s.ldm
+}
+
+// This function will close client and route listeners
+// then close clients at some interval to avoid a reconnecting storm.
+// Note that the routes are closed before so that when a client is
+// disconnected and reconnects to another server, that server does
+// not send to this server a route SUB.
+func (s *Server) lameDuckMode() {
```
Comment (ripienaar) [inline server/server.go]: These 2 functions would be useful as well to people who embed it in other software. 

Or some other interface that achieves the lame duck shutdown like an optional flag to Shutdown()

File: server/server.go (hunk 8 above)
Comment (kozlovic) [reply server/server.go]: I have some pending changes, but still did not implement that. The more I think about it, how is that better? Say there are 1000 clients to close within 30 seconds. There is timer firing every 1 sec, what do you expect the timer callback to do? Close a certain number of clients in a row then reset the timer to fire in 1 sec? I am not sure that this is good because there will still be a bunch of reconnect at once, as opposed to spread the closing of the clients (still may require to close more than one in a row if there is many more clients that the number of ms left).

Comment (kozlovic) [reply server/server.go]: That is, the pause of 1sec between each interval is helping I guess, but not as much as making sure that - if we can - we spread the closing, I think.

Comment (derekcollison) [reply server/server.go]: I think you are right. Even though we expect clusters of a size that all of the batch clients connect to not the same machine, spreading it out a bit even within a batch makes more sense. We should pick a timer interval that allows us 1 client close per tick (as best we can).

File: server/server.go (hunk 9 above)
Comment (derekcollison) [reply server/server.go]: @ripienaar What functionality would you want to happen here from a client perspective? Maybe callback when state transitioning from lame duck to shutting down?

Comment (ripienaar) [reply server/server.go]: @derekcollison from a client perspective. Maybe the jet lag talking but I am not sure I follow the question

As I understand this code it’s a shutdown like any other shutdown it just takes a while since it closes clients down over time nothing more? Client shouldn’t care and from the point of view of the code embedding it it should just be like Shutdown(). So something like gnatsd.Shutdown(server.LameDuckShutdown()) would do what I need quite nicely.

**Comment by ripienaar** [comment]
In my case I embed a server and it’s routed with others - so same use case and issues hold for me. Clients are remote.

**Comment by derekcollison** [comment]
ok gotcha.

File: server/server.go (hunk 8 above)
Comment (kozlovic) [reply server/server.go]: Honestly, using a timer here makes the code way more complex than the original in the PR (although I would change the 10ms sleep between batches. Are you sure about the timer approach?

File: server/server.go [hunk 10]
```diff
@@ -594,6 +599,15 @@ func (s *Server) AcceptLoop(clr chan struct{}) {
...
 		conn, err := l.Accept()
 		if err != nil {
+			if s.isLameDuckMode() {
+				// Signal that we are not accepting new clients
+				s.ldmCh <- true
+				// We need to wait for lameDuckMode() to notify us that it
+				// is done with closing clients. If we return before that,
+				// the s.Start() function would return and process would exit.
+				<-s.ldmDoneCh
+				return
```
[the remark this answers is not in the document]
Comment (kozlovic) [reply server/server.go]: With the latest changes, did not need the `ldmDone channel`.

**Comment by derekcollison** [comment]
LGTM
